ethers/providers/jsonrpc/subscriptions.nim

```diff
@@ -152,47 +150,60 @@ proc new*(_: type JsonRpcSubscriptions,
 
   let subscriptions = PollingSubscriptions(client: client)
 
-  proc getChanges(originalId: JsonNode): Future[JsonNode] {.async.} =
+  proc resubscribe(id: JsonNode) {.async: (raises: [CancelledError]).} =
     try:
-      let mappedId = subscriptions.subscriptionMapping[originalId]
-      let changes = await subscriptions.client.eth_getFilterChanges(mappedId)
-      if changes.kind == JNull:
-        return newJArray()
-      elif changes.kind != JArray:
-        raise newException(SubscriptionError,
-          "HTTP polling: unexpected value returned from eth_getFilterChanges." &
-          " Expected: JArray, got: " & $changes.kind)
-      return changes
-    except CancelledError as e:
-      raise e
-    except CatchableError as e:
-      if "filter not found" in e.msg:
-        var newId: JsonNode
-        # Log filters are stored in logFilters, block filters are not persisted
-        # there is they do not need any specific data for their recreation.
-        # We use this to determine if the filter was log or block filter here.
-        if subscriptions.logFilters.hasKey(originalId):
-          let filter = subscriptions.logFilters[originalId]
-          newId = await subscriptions.client.eth_newFilter(filter)
-        else:
-          newId = await subscriptions.client.eth_newBlockFilter()
-        subscriptions.subscriptionMapping[originalId] = newId
-        return await getChanges(originalId)
+      var newId: JsonNode
+      # Log filters are stored in logFilters, block filters are not persisted
+      # there is they do not need any specific data for their recreation.
+      # We use this to determine if the filter was log or block filter here.
+      if subscriptions.logFilters.hasKey(id):
+        let filter = subscriptions.logFilters[id]
+        newId = await subscriptions.client.eth_newFilter(filter)
       else:
-        raise e
+        newId = await subscriptions.client.eth_newBlockFilter()
+      subscriptions.subscriptionMapping[id] = newId
+    except CancelledError as error:
+      raise error
+    except CatchableError:
+      # there's nothing further we can do here
+      discard
 
-  proc poll(id: JsonNode) {.async.} =
+  proc getChanges(id: JsonNode): Future[JsonNode] {.async: (raises: [CancelledError]).} =
+    if mappedId =? subscriptions.subscriptionMapping.?[id]:
+      try:
+        let changes = await subscriptions.client.eth_getFilterChanges(mappedId)
+        if changes.kind == JArray:
+          return changes
+      except JsonRpcError:
+        await resubscribe(id)
```

Comment:
My only reservation here is that we may be resubscribing in too many situations. In the case of a disconnection, we don't really need to resubscribe as the subscription should still be on the server. Imagine there's a disconnection for 1 minute. At four second polling intervals, that'll be 15 new subscriptions. It may contribute to bloated change sets...

The main idea before was to keep the polling loop from silently crashing, by restarting the polling loop when it errored. However, you have covered this by asyncSpawning the polling loop and ensuring there are no exceptions leaked from it.

Reply:
> Imagine there's a disconnection for 1 minute. At four second polling intervals, that'll be 15 new subscriptions.

Not if we're disconnected, then the re-subscriptions will also fail.

But I agree, this could be more elegant. Ideally I'd like to move to how it's done in ethers.js. If I understand it correctly, they only use a single filter to watch for new blocks, and they parse the block logs to see if there are any subscriptions in memory that need to be triggered.

Reply:
`JsonRpcError` is a base-level error, of which all other types of errors in nim-json-rpc are derived. If we re-subscribe on all errors of this type, we may be doing ourselves a disservice and re-subscribing unnecessarily. For example, the error types `ErrorResponse`, `InvalidRequest`, `ApplicationError`, etc are all derived from `JsonRpcError`. This means that we may have encountered an error that does not indicate the filters were discarded or not found server-side. In my opinion, we don't necessarily NEED to resubscribe unless we encountered a `filter not found` error, (guess: would be a type of `ErrorResponse`). So, I think maybe we should re-introduce this condition, eg:

```nim
proc getChanges() =
  try:
    getFilterChanges
  except ErrorResponse as e:
    if "filter not found" in e.msg:
      await resubscribe(id)
  except CancelledError as e:
    raise e
  except CatchableError as e:
    discard
```
This would still require the additional work to ensure we don't miss events between losing the subscription and resubscription.

Reply:
I checked, and the error type in the RPC client would likely be a `JsonRpcError` instead of `ErrorResponse`, so the above block would be:
```nim
proc getChanges() =
  try:
    getFilterChanges
  except JsonRpcError as e:
    if "filter not found" in e.msg:
      await resubscribe(id)
  except CancelledError as e:
    raise e
  except CatchableError as e:
    discard
```

Reply:
Yes, I actually tested all of the subtypes of JsonRpcError, but in this case we only get the base error type out of it.

I did not want to do matching on the error message string, because there is no guarantee that all node types return this exact text.